### DISCUSSION TITLE
"Endless" playback: introducing a new, "Sustained" state

### DISCUSSION THREAD

**Discussion by saintmatthieu** [audacity/audacity#7311] [Near future]
## Background

We are considering modifying the playback behavior to improve the user experience. In short, we would like to give a chance for effects with a tail (echo, reverb, etc.) to continue output audio after the playhead stops.

## TL;DR
Here are the major points:
* One only gets to the Sustained state from the Playing state **and** by pressing STOP (or the space shortcut), which also resets the playhead.
* Sustained state is a seemingly Stopped state, except that STOP and EDIT bring back to the Stopped state,

https://github.com/user-attachments/assets/e824833d-009a-487d-b600-4089cd57bf33

https://github.com/user-attachments/assets/ad53683f-76df-41a4-b492-a5de56fa3e48

* In Recording state, STOP brings to the Stopped state (not the Sustained state).

## New things to consider:
* In Sustained state, there isn't an existing shortcut to stop playback, only the stop button. @LWinterberg FYI
* We should probably adapt the state indication in the UI, change it to "Sustaining" (?) when appropriate:
![image](https://github.com/user-attachments/assets/4b7c6406-5a7a-45ae-938d-5e3518ad3105)

## State diagram

There are five events

-  PAUSE
-  PLAY
-  STOP
-  RECORD
-  EDIT

The EDIT event broadly denotes all events that are currently prohibited during playback, such as edits of the track panel, applying effects, creating a new project, etc.

We would have the current six states, plus a new one, Sustained.

-  Stopped
-  Playing
-  Recording
-  Stopped Paused
-  Playing Paused
-  Recording Paused
-  Sustained (new)

In Sustained state, the playhead isn't moving, but the audio continues to play, giving a chance to effects with a tail to continue output audio.

Here's a state diagram of what the playback would be like:

```mermaid
stateDiagram-v2
    [*] --> Stopped

   RecordingPaused: Recording Paused
   StopPaused: Stopped Paused
   PlayingPaused: Playing Paused

   StopPaused --> Stopped: PAUSE
   StopPaused --> Playing: PLAY
   StopPaused --> RecordingPaused: RECORD
   note right of StopPaused
       STOP disabled
       EDIT doesn't change state
   end note

   RecordingPaused --> Recording: PAUSE
   RecordingPaused --> Stopped: STOP<br/>EDIT
   note right of RecordingPaused
       PLAY disabled
       RECORD ineffective
   end note

   PlayingPaused --> Playing: PAUSE<br/>PLAY
   PlayingPaused --> Stopped: STOP<br/>EDIT
   PlayingPaused --> Recording: RECORD

    Stopped --> Playing: PLAY
    Stopped --> Recording: RECORD
    Stopped --> StopPaused: PAUSE
    note right of Stopped
        STOP disabled
        EDIT doesn't change state
    end note

    Playing --> PlayingPaused: PAUSE
    Playing --> Sustained: STOP
    note right of Playing
        PLAY resets playhead
        RECORD & EDIT disabled
    end note

   Recording --> RecordingPaused: PAUSE
   Recording --> Stopped: STOP
   note right of Recording
       PLAY & EDIT disabled
       RECORD ineffective
   end note

    Sustained --> Playing: PLAY
    Sustained --> Stopped: STOP<br/>EDIT
   Sustained --> Recording: RECORD
   Sustained --> StopPaused: PAUSE

```

The only difference to the current state diagram is that the Playing state now transitions to the Sustained state when the user stops the playback. Then, in Sustained state,
* PLAY brings back to the Playing state,
* STOP and EDIT bring to the Stopped state,
* RECORD brings to the Recording state,
* PAUSE brings to the Stopped Paused state.

## PoC branch
https://github.com/saintmatthieu/audacity/tree/2617-endless-playback

**Comment by LWinterberg** [comment]
NB: In the STOPPED state currently has a direct path to the PAUSED state - you can do STOPPED -> PAUSE -> RECORD to arm Audacity's recording and seeing the metering and such, and saving you the step of doing a test recording.

**Reply by saintmatthieu** [reply]
A revelation, thank you! Diagram updated.

**Comment by saintmatthieu** [comment]
Endless playback ticket: https://github.com/audacity/audacity/issues/2617

**Comment by LWinterberg** [comment]
> In Sustained state, there isn't an existing shortcut to stop playback, only the stop button. @LWinterberg FYI

Sending any of the stop commands (Play/Stop, Play once/Stop, Play/Stop and set cursor, Stop) should put playback into the Sustained state. Sending the Stop command again should put it in the Stopped state. Sending the Stop command additional times should be enabled, but a NOOP.

**Comment by DavidBailes** [comment]
Can you just clarify: If playback reaches the end of a time selection or the end of the audio in a project, will the playback state change to stopped, rather than sustained?

One or two issues come immediately to mind:
1. Using spacebar to start/stop playback. There will now have to be an extra press of spacebar if a user wants to stop and then restart playback. I don't think this will be very intuitive.
2. Is "sustained playback" when the playback cursor is not moving slightly strange?
3. The transport status in the status bar. If the transport status bar is going to be "sustained", is it going to be at all obvious to the user what this means?

**Reply by LWinterberg** [reply]
> Using spacebar to start/stop playback. There will now have to be an extra press of spacebar if a user wants to stop and then restart playback. I don't think this will be very intuitive.

The Sustained state for most intents and purposes is going to be the same as the Stopped state. If you have no realtime effects which produce a large tail (so most things except reverb/delay), you'll hear no difference. For the play/stop buttons, they would not become tri-state, but rather remain 2-state (with the second state being sustain)

> Is "sustained playback" when the playback cursor is not moving slightly strange?

No new data is being sent to the playback, if you go into sustain just before a new impulse, the impulse will never play. Only reverb on the existing data will be allowed to ring out. This is how it works in most DAWs. 

> The transport status in the status bar. If the transport status bar is going to be "sustained", is it going to be at all obvious to the user what this means?

Wording is still up for debate. We might not even expose the state to the user.

**Comment by DavidBailes** [comment]
Another issue or two:
1. The commands Play/Stop etc will  no longer do what they say they will do. Will they be renamed Play/Sustained/Stop etc?
2. Will the Stop button in the transport toolbar renamed Sustained?

**Reply by LWinterberg** [reply]
1. no, they will stay 2-state
2. no.

**Reply by DavidBailes** [reply]
> 1. no, they will stay 2-state

If the commands such as Play/Stop are going to remain two state, then wouldn't it be a lot more use friendly if when the playback state was sustained, the Play/Stop command started playback, rather than changing playback to stopped state. Then the user could press spacebar to stop, then spacebar to start playback again, just as they can at the moment. I really don't think having to press spacebar an additional time to restart playback is all intuitive.

**Reply by LWinterberg** [reply]
Again: They will remain 2-state. 

1. press on Play/Stop starts playing
2. press on Play/Stop stops playback and lets realtime effects ring out
3. press on Play/Stop ends any ringing out that may still be present and starts playing
4. press on Play/Stop stops playback and lets realtime effects ring out

On a technical level, these commands will toggle between Playing and Sustained. As far as the user is concerned, when stopping, realtime effects are now able to ring out.

**Reply by DavidBailes** [reply]
Thanks for the clarification. I think that my confusion arose because the effect of the Play/Stop commands in not included in @saintmatthieu 's state diagram at the start of this issue. Maybe they should be included.

**Comment by saintmatthieu** [comment]
I think this is the part we are talking about ?
![image](https://github.com/user-attachments/assets/89064b2b-57c5-46fe-b593-5ec8b59ff781)
(Looking at it I realized there was a mistake, a transition going from Playing to Stopped triggered by STOP. I just removed it.)
Could you describe what you think is missing ?

**Reply by DavidBailes** [reply]
> Could you describe what you think is missing ?

My brain working. There isn't a problem!
